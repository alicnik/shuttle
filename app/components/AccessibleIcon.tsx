import * as AccessibleIconComponent from '@radix-ui/react-accessible-icon';

export const AccessibleIcon = AccessibleIconComponent.Root;
