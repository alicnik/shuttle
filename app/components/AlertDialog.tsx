import {
  AlertDialog as AlertDialogComponent,
  AlertDialogAction,
  AlertDialogCancel,
  AlertDialogContent,
  AlertDialogDescription,
  AlertDialogFooter,
  AlertDialogHeader,
  AlertDialogTitle,
} from '~/components/ui/alert-dialog';

type AlertDialogProps = React.PropsWithChildren<{
  title: string;
  description: string;
  onConfirm: () => void;
}>;

export function AlertDialog({
  children,
  title,
  description,
  onConfirm,
}: AlertDialogProps) {
  return (
    <AlertDialogComponent>
      {children}
      <AlertDialogContent onCloseAutoFocus={(e) => e.preventDefault()}>
        <AlertDialogHeader>
          <AlertDialogTitle>{title}</AlertDialogTitle>
          <AlertDialogDescription>{description}</AlertDialogDescription>
        </AlertDialogHeader>
        <AlertDialogFooter>
          <AlertDialogCancel>Cancel</AlertDialogCancel>
          <AlertDialogAction onClick={onConfirm}>Continue</AlertDialogAction>
        </AlertDialogFooter>
      </AlertDialogContent>
    </AlertDialogComponent>
  );
}
