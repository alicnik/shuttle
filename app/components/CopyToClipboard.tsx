import { ClipboardCopyIcon } from '@radix-ui/react-icons';
import { useToast } from './ui';
import { Tooltip } from './Tooltip';
import type { TooltipProps } from './Tooltip';

interface CopyToClipboardProps {
  copyText: string;
  successMessage: string;
  tooltipOptions: Omit<TooltipProps, 'children'>;
}

export function CopyToClipboard({
  copyText,
  successMessage,
  tooltipOptions: { content, side = 'top', sideOffset = 0 },
}: CopyToClipboardProps) {
  const { toast } = useToast();

  return (
    <Tooltip content={content} side={side} sideOffset={sideOffset}>
      <ClipboardCopyIcon
        className="translate-y-[3px] cursor-pointer"
        onClick={async () => {
          await navigator.clipboard.writeText(copyText);
          toast({ description: successMessage });
        }}
      />
    </Tooltip>
  );
}
