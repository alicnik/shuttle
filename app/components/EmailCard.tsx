import type { Email } from '@prisma/client';
import { Checkbox } from '~/components/ui/checkbox';
import { useNavigation, useSubmit } from '@remix-run/react';
import { formatRelative } from 'date-fns';

interface EmailCardProps {
  email: Email;
  selected: string[];
  setSelected: React.Dispatch<React.SetStateAction<string[]>>;
  preview: string | null;
  setPreview: React.Dispatch<React.SetStateAction<string | null>>;
}

export function EmailCard({
  email,
  selected,
  setSelected,
  preview,
  setPreview,
}: EmailCardProps) {
  const submit = useSubmit();
  const navigation = useNavigation();

  if (navigation.state === 'submitting') {
    console.log(Object.fromEntries(navigation.formData || []));
  }

  const isOptimisticRead =
    ['submitting', 'loading'].includes(navigation.state) &&
    navigation.formData?.get('_action') === 'markRead' &&
    navigation.formData?.get('emailId') === email.id;

  const isOptimisticUnread =
    ['submitting', 'loading'].includes(navigation.state) &&
    navigation.formData?.get('_action') === 'markSelectedUnread' &&
    String(navigation.formData?.get('selected')).split(',').includes(email.id);

  const isSelected = selected.includes(email.id);
  const isPreview = preview === email.id;

  const from = email.from.startsWith('<')
    ? email.from.slice(1, -1)
    : email.from.split('<')[0].trim();

  return (
    <div
      data-selected={isSelected}
      data-in-view={isPreview}
      className="mb-4 flex gap-2 rounded-md p-2 hover:bg-zinc-800 data-[in-view=true]:bg-zinc-800 data-[selected=true]:bg-zinc-800"
    >
      <Checkbox
        className="mt-[2px]"
        checked={isSelected || isPreview}
        onCheckedChange={(checked) => {
          setSelected((selected) => {
            if (checked) {
              return [...selected, email.id];
            }

            if (isPreview) {
              setPreview(null);
            }
            return selected.filter((id) => id !== email.id);
          });
        }}
      />
      <div
        data-read={!isOptimisticUnread && (email.read || isOptimisticRead)}
        className="cursor-pointer text-left data-[read=true]:opacity-50"
        onClick={() =>
          setPreview((current) => {
            if (current) {
              submit(
                { emailId: current, _action: 'markRead' },
                { method: 'post' }
              );
            }
            return email.id;
          })
        }
      >
        <div className="mb-1 flex items-center ">
          <p className="mr-12 w-max text-sm font-bold">{from}</p>
          <p className="ml-auto w-max text-xs">
            {formatRelative(email.createdAt, new Date())}
          </p>
        </div>
        <p className="mb-1 text-sm">{email.subject}</p>
        <p className="line-clamp-2 text-xs">{email.text}</p>
      </div>
    </div>
  );
}
