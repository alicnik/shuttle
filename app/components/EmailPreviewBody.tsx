import clsx from 'clsx';
import type { Email } from '@prisma/client';
import * as React from 'react';

export function EmailPreviewBody({ html }: { html?: Email['html'] }) {
  const htmlWithBaseTag = React.useMemo(() => {
    if (!html) return '';

    // This ensures links open in a new tab and not in the iframe
    const baseTag = '<base target="_blank">';
    const headTag = '<head>';
    const headTagIndex = html.indexOf(headTag);

    if (headTagIndex === -1) {
      return html;
    }

    return (
      html.slice(0, headTagIndex + headTag.length) +
      baseTag +
      html.slice(headTagIndex + headTag.length)
    );
  }, [html]);

  return (
    <iframe
      data-testid="email-preview-body"
      className={clsx('email-preview', 'h-full w-full')}
      srcDoc={htmlWithBaseTag}
      title="Email preview"
    />
  );
}
