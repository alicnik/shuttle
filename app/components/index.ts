export * from './AlertDialog';
export * from './EmailCard';
export * from './EmailPreviewBody';
export * from './EmailPreviewHeader';
export * from './Tooltip';
