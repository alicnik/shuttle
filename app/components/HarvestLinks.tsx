import * as React from 'react';
import {
  Button,
  Dialog,
  DialogContent,
  DialogDescription,
  DialogHeader,
  DialogTitle,
  DialogTrigger,
  useToast,
} from './ui';
import { Tooltip } from './Tooltip';
import { Link2Icon } from '@radix-ui/react-icons';
import { LINK_COPY_SUCCESS_MESSAGE } from '~/lib';
import type { Email } from '@prisma/client';
import { AccessibleIcon } from './AccessibleIcon';

interface HarvestLinksProps {
  email: Email;
}

export function HarvestLinks({ email }: HarvestLinksProps) {
  const { toast } = useToast();

  const links = React.useMemo(() => {
    /* c8 ignore next 3 */
    if (typeof window === 'undefined') {
      return [];
    }
    const parser = new DOMParser();
    const doc = parser.parseFromString(email.html, 'text/html');
    const links = Array.from(doc.querySelectorAll('a'));
    const mapped = links.map((link) => {
      const linkDoc = parser.parseFromString(link.innerHTML, 'text/html');
      const elements = linkDoc.querySelectorAll('*');
      elements.forEach((element) => {
        element.removeAttribute('style');
        element.removeAttribute('width');
        element.removeAttribute('height');
        if (element.tagName === 'IMG') {
          element.setAttribute('style', 'max-width: 80%;');
        }
      });
      link.innerHTML = linkDoc.body.innerHTML;
      return link;
    });
    return mapped;
  }, [email.html]);

  return (
    <Dialog>
      <Tooltip content="Harvest links">
        <DialogTrigger asChild>
          <Button
            variant="ghost"
            className="hover:bg-zinc-300"
            type="button"
            size="icon"
          >
            <AccessibleIcon label="Harvest links from email">
              <Link2Icon />
            </AccessibleIcon>
          </Button>
        </DialogTrigger>
      </Tooltip>
      <DialogContent onCloseAutoFocus={(e) => e.preventDefault()}>
        <DialogHeader>
          <DialogTitle>Found the following links in the email</DialogTitle>
          <DialogDescription>
            Click a link to copy it to your clipboard
          </DialogDescription>
          {links.length > 0 ? (
            <ul className="grid auto-rows-fr grid-cols-2 place-items-center gap-x-2 gap-y-8 pt-8">
              {links.map((link) => {
                return (
                  <Tooltip key={link.href} content={link.href} sideOffset={4}>
                    <li
                      className="flex cursor-pointer justify-center"
                      dangerouslySetInnerHTML={{
                        __html: link.innerHTML,
                      }}
                      onClick={async () => {
                        await navigator.clipboard.writeText(link.href);
                        toast({
                          description: LINK_COPY_SUCCESS_MESSAGE,
                        });
                      }}
                    />
                  </Tooltip>
                );
              })}
            </ul>
          ) : (
            <DialogDescription>
              No links were found in the email
            </DialogDescription>
          )}
        </DialogHeader>
      </DialogContent>
    </Dialog>
  );
}
