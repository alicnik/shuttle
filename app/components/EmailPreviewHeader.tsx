import * as React from 'react';
import { Form, Link, useSearchParams, useSubmit } from '@remix-run/react';
import { format } from 'date-fns';
import { Tooltip } from './Tooltip';
import { Button, Separator } from './ui';
import type { Email } from '@prisma/client';
import {
  EnvelopeClosedIcon,
  ExternalLinkIcon,
  TrashIcon,
} from '@radix-ui/react-icons';
import { AlertDialog, AlertDialogTrigger } from './AlertDialog';
import { HarvestLinks } from './HarvestLinks';
import { AccessibleIcon } from './AccessibleIcon';

interface EmailPreviewHeaderProps {
  email: Email;
  setSelected: React.Dispatch<React.SetStateAction<string[]>>;
}

export function EmailPreviewHeader({
  email,
  setSelected,
}: EmailPreviewHeaderProps) {
  const submit = useSubmit();
  const [, setSearchParams] = useSearchParams();

  return (
    <div className="max-h-48 overflow-hidden p-4 text-zinc-900">
      <h3 className="truncate">
        <span className="font-bold">From:</span> {email.from}
      </h3>
      <h3 className="truncate">
        <span className="font-bold">To:</span> {`${email.userId}@shuttle.email`}
      </h3>
      <h3 className="truncate">
        <span className="font-bold">Subject:</span> {email.subject}
      </h3>
      <h3 className="mb-4 truncate">
        <span className="font-bold">Received:</span>{' '}
        {format(email.createdAt, 'cccc, do MMMM yyyy @ hh:mm a')}
      </h3>
      <Form method="post" className="mb-4 flex gap-2">
        <input type="hidden" name="selected" value={email.id} />
        <Tooltip content="Mark as unread">
          <Button
            variant="ghost"
            size="icon"
            name="_action"
            value="markSelectedUnread"
            className="hover:bg-zinc-300"
          >
            <AccessibleIcon label="Mark as unread">
              <EnvelopeClosedIcon />
            </AccessibleIcon>
          </Button>
        </Tooltip>
        <AlertDialog
          title="Are you sure?"
          description="This action cannot be undone."
          onConfirm={() => {
            setSearchParams((params) => {
              params.delete('preview');
              return params;
            });
            setSelected((selected) => selected.filter((id) => id !== email.id));
            submit(
              { selected: email.id, _action: 'deleteSelected' },
              { method: 'post' }
            );
          }}
        >
          <Tooltip content="Delete">
            <AlertDialogTrigger asChild>
              <Button
                name="_action"
                value="deleteSelected"
                variant="ghost"
                size="icon"
                className="hover:bg-destructive"
              >
                <AccessibleIcon label="Delete">
                  <TrashIcon />
                </AccessibleIcon>
              </Button>
            </AlertDialogTrigger>
          </Tooltip>
        </AlertDialog>
        <Tooltip content="Open in new window">
          <Link to={`/${email.userId}/${email.id}`} target="_blank">
            <Button
              type="button"
              variant="ghost"
              className="hover:bg-zinc-300"
              size="icon"
              onClick={() => {
                submit(
                  { emailId: email.id, _action: 'markRead' },
                  { method: 'post' }
                );
              }}
            >
              <AccessibleIcon label="Open in new window">
                <ExternalLinkIcon />
              </AccessibleIcon>
            </Button>
          </Link>
        </Tooltip>
        <HarvestLinks email={email} />
      </Form>
      <Separator />
    </div>
  );
}
