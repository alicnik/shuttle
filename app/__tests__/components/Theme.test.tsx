import { render } from '@testing-library/react';
import { ThemeScript } from '~/components';

describe('Theme', () => {
  it('should match the snapshot', () => {
    const { container } = render(<ThemeScript />);

    expect(container).toMatchSnapshot();
  });
});
