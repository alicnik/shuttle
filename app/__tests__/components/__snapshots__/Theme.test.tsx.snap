// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Theme > should match the snapshot 1`] = `
<div>
  <script>
    
      const prefersDarkMode = window.matchMedia("(prefers-color-scheme: dark)").matches
      if (prefersDarkMode) {
        document.documentElement.classList.add('dark')
      }
    
  </script>
</div>
`;
